src/recovery/hystart.rs

```diff
@@ -88,93 +116,107 @@ impl Hystart {
-        if self.lss_start_time().is_none() {
-            if let Some(current_round_min_rtt) = self.current_round_min_rtt {
-                self.current_round_min_rtt =
-                    Some(cmp::min(current_round_min_rtt, rtt));
-            } else {
-                self.current_round_min_rtt = Some(rtt);
-            }
+        if !(self.enabled && epoch == packet::EPOCH_APPLICATION) {
+            return false;
+        }
+
+        self.current_round_min_rtt = cmp::min(self.current_round_min_rtt, rtt);
 
-            self.rtt_sample_count += 1;
+        self.rtt_sample_count += 1;
 
+        // Slow Start.
+        if self.css_start_time().is_none() {
             if cwnd >= (LOW_CWND * max_datagram_size) &&
-                self.rtt_sample_count >= N_RTT_SAMPLE &&
-                self.current_round_min_rtt.is_some() &&
-                self.last_round_min_rtt.is_some()
+                self.rtt_sample_count >= N_RTT_SAMPLE
             {
+                self.rtt_sample_count = 0;
+
                 // clamp(min_rtt_thresh, last_round_min_rtt/8,
                 // max_rtt_thresh)
-                let rtt_thresh = cmp::max(
-                    self.last_round_min_rtt.unwrap() / 8,
-                    MIN_RTT_THRESH,
-                );
+                let rtt_thresh =
+                    cmp::max(self.last_round_min_rtt / 8, MIN_RTT_THRESH);
                 let rtt_thresh = cmp::min(rtt_thresh, MAX_RTT_THRESH);
 
-                // Check if we can exit to LSS.
-                if self.current_round_min_rtt.unwrap() >=
-                    (self.last_round_min_rtt.unwrap() + rtt_thresh)
+                // Check if we can exit to CSS.
+                if self.current_round_min_rtt >=
+                    self.last_round_min_rtt.saturating_add(rtt_thresh)
                 {
-                    self.lss_start_time = Some(now);
+                    self.css_baseline_min_rtt = self.current_round_min_rtt;
+                    self.css_start_time = Some(now);
                 }
             }
+        } else {
+            // Conservative Slow Start.
+            if self.rtt_sample_count >= N_RTT_SAMPLE {
+                self.rtt_sample_count = 0;
+
+                if self.current_round_min_rtt < self.css_baseline_min_rtt {
+                    self.css_baseline_min_rtt = Duration::MAX;
 
-            // Check if we reached the end of the round.
-            if let Some(end_pkt_num) = self.window_end {
-                if packet.pkt_num >= end_pkt_num {
-                    // Start of a new round.
-                    self.window_end = None;
+                    // Back to Slow Start.
+                    self.css_start_time = None;
+                    self.css_round_count = 0;
                 }
             }
         }
 
-        self.lss_start_time.is_some()
+        // Check if we reached the end of the round.
+        if let Some(end_pkt_num) = self.window_end {
+            if packet.pkt_num >= end_pkt_num {
```

Comment:
What if a packet later than `window_end` is acked before `window_end` (e.g. due to re-ordering)? Is it ok to end the round anyway? The draft doesn't seem to talk about this.

Reply:
The algorithm in the draft is based on TCP, so the definition of the round is different in QUIC. When reordering is enough to trigger a congestion event we will end SS/CSS any way. If this is tolerable within reordering threshold, I don't have too much concern about it.